README.md
# MMM-AlarmClock [![GitHub license](https://img.shields.io/badge/license-MIT-blue.svg?style=flat)](https://raw.githubusercontent.com/fewieden/MMM-AlarmClock/master/LICENSE) [![Build Status](https://travis-ci.org/fewieden/MMM-AlarmClock.svg?branch=master)](https://travis-ci.org/fewieden/MMM-AlarmClock) [![Code Climate](https://codeclimate.com/github/fewieden/MMM-AlarmClock/badges/gpa.svg?style=flat)](https://codeclimate.com/github/fewieden/MMM-AlarmClock) [![Known Vulnerabilities](https://snyk.io/test/github/fewieden/mmm-alarmclock/badge.svg)](https://snyk.io/test/github/fewieden/mmm-alarmclock)

Alarm Clock Module for MagicMirror<sup>2</sup>

## Example

![Active alarm](.github/example.jpg)   ![Scheduled alarm](.github/example2.jpg)

## Dependencies

* An installation of [MagicMirror<sup>2</sup>](https://github.com/MichMich/MagicMirror)

## Installation

* Clone this repo into `~/MagicMirror/modules` directory.
* Configure your `~/MagicMirror/config/config.js`:

```js
{
    module: 'MMM-AlarmClock',
    position: 'top_right',
    config: {
        alarms: [
            {time: "18:30", days: [2,4], title: "Soccer", message: "Get ready for soccer training!", sound: "alarm.mp3"},
            ...
        ],
        ...
    }
}
```

## Config Options

| **Option** | **Default** | **Description** |
| --- | --- | --- |
| `alarms` | `REQUIRED` | An Array with all your alarms as objects. Those objects need to have the properties -> time: 24h format, days: Array of all days the alarm should be fired (0 = Sunday, 6 = Saturday), title, message, and sound. If sound is not defined, alarm will be fired with module defined/default alarm sound. |
| `sound` | `'alarm.mp3'` | Name and extension of your alarm sound. File needs to be placed in `~/MagicMirror/modules/MMM-AlarmClock/sounds`. Standard files are `alarm.mp3` and `blackforest.mp3`.  Alternatively specify a web stream `http` or `https`. |
| `volume` | `1.0` | The volume of the alarm sound in a range from `0.0` to `1.0` |
| `touch` | `false` | If you are using a touch screen device you need to press a button to disable an alarm.. |
| `format` | `'ddd, h:mmA'` | In which format the alarm in the header should be displayed. [All Options](http://momentjs.com/docs/#/displaying/format/) |
| `timer` | `60000` (1 min) | How long the alarm should ring for non touch screen or without interaction on touch screen devices. |
| `fade` | `false` | Set to enable a gradual fade-in of the alarm sound. |
| `fadeTimer` | `60 * 1000` (1 min) | How long to fade into the alarm before `volume` is set. |
| `fadeStep` | `.005` (.5%) | Increase the volume this percent amount each second until `fadeTimer` is reached. |
| `popup` | `true` | Flag to show alert popup or not. |

## Alarm Options

| **Option**| **Description** |
| --- | --- |
| `time` | Time at which the alarm should sound. |
| `days` | Days of the week. <br/>``0 => Sunday, 1 => Monday, 2 => Tuesday, 3 => Wednesday, 4 => Thursday, 5 => Friday, 6 => Saturday``<br/><br/>**Example :** ``[1, 2, 3, 4, 5], // From Monday to Friday.`` |
| `title` |  Title that will be displayed in the alert. |
| `message` | Message of the alarm |
| `sound` | Name or the url of the mp3 file. By default, the sound of the config will be use if this option is empty. |
| `timer` | Timer when the alarm will end. By default, the timer of the configuration will be used. |

## Alarm Sounds

There are already two alarm sounds:

* [alarm.mp3](http://www.orangefreesounds.com/mp3-alarm-clock/) | From Alexander licensed under [CC BY-NC 4.0](https://creativecommons.org/licenses/by-nc/4.0/)
* [blackforest.mp3](http://www.orangefreesounds.com/coo-coo-clock-sound/) | From Alexander licensed under [CC BY-NC 4.0](https://creativecommons.org/licenses/by-nc/4.0/)

## Developer

* `npm run lint` - Lints JS and CSS files.
* `npm run docs` - Generates documentation.


CHANGELOG.md
# MMM-AlarmClock Changelog

## [Unreleased]

### Fixed

### Added

### Changed

### Removed

## [2.0.2]

### Fixed

* Timing of alert click handler registration

## [2.0.1]

### Fixed

* Fading of alert sound.

### Added

* Korean translations

## [2.0.0]

### Added

* Config option `popup` to avoid alert popup.
* Nunjuck templates
* Notification handler `STOP_ALARM` to stop an alarm from another module.

### Changed

* Config files
* Removed documentation on doclets.io (unmaintained).
* eslint recommended instead of airbnb ruleset.

## [1.1.1]

### Added

* Add a new alarm option : timer
* Add a .gitignore in sounds folder.

## [1.1.0]

### Added

* Volume fade in (see config options fade, fadeTimer and fadeStep).
* French translations
* Indonesian translations
* Documentation
* [Doclets.io](https://doclets.io/fewieden/MMM-AlarmClock/master) integration
* Contributing guidelines
* Issue template
* Pull request template
* gitignore

## [1.0.0]

Initial version


MMM-AlarmClock.js
/**
 * @file MMM-AlarmClock.js
 *
 * @author fewieden
 * @license MIT
 *
 * @see  https://github.com/fewieden/MMM-AlarmClock
 */

/* global Module Log moment config MM */

/**
 * @external Module
 * @see https://github.com/MichMich/MagicMirror/blob/master/js/module.js
 */

/**
 * @external Log
 * @see https://github.com/MichMich/MagicMirror/blob/master/js/logger.js
 */

/**
 * @external moment
 * @see https://www.npmjs.com/package/moment
 */

/**
 * @external config
 * @see https://github.com/MichMich/MagicMirror/blob/master/config/config.js.sample
 */

/**
 * @external MM
 * @see https://github.com/MichMich/MagicMirror/blob/master/js/main.js
 */

/**
 * @module MMM-AlarmClock
 * @description Frontend for the module to display data.
 *
 * @requires external:Module
 * @requires external:Log
 * @requires external:moment
 * @requires external:config
 * @requires external:MM
 */
Module.register('MMM-AlarmClock', {

    /** @member {?Object} next - Contains the next alarm object. */
    next: null,
    /** @member {boolean} alarmFired - Flag that indicates if there is a alarm firing right now. */
    alarmFired: false,

    /** @member {?Timeout} timer - Fires resetAlarmClock */
    timer: null,
    /** @member {?Interval} fadeInterval - Fades in the alarm volume. */
    fadeInterval: null,

    /**
     * @member {Object} defaults - Defines the default config values.
     * @property {string} sound - Alarm sound file.
     * @property {boolean} touch - Flag to enable touch support.
     * @property {number} volume - Alarm sound volume.
     * @property {string} format - Alarm datetime format to display in UI.
     * @property {int} timer - Alarm sound duration.
     * @property {boolean} fade - Flag to fade in alarm sound volume.
     * @property {int} fadeTimer - Fade in duration.
     * @property {number} fadeStep - Fade in volume steps.
     * @property {boolean} popup - Flag to show alert popup or not.
     */
    defaults: {
        sound: 'alarm.mp3',
        touch: false,
        volume: 1.0,
        format: 'ddd, h:mmA',
        timer: 60 * 1000,
        fade: false,
        fadeTimer: 60 * 1000,
        fadeStep: 0.005,
        popup: true
    },

    /**
     * @function getStyles
     * @description Style dependencies for this module.
     * @override
     *
     * @returns {string[]} List of the style dependency filepaths.
     */
    getStyles() {
        return ['font-awesome.css', 'MMM-AlarmClock.css'];
    },

    /**
     * @function getScripts
     * @description Script dependencies for this module.
     * @override
     *
     * @returns {string[]} List of the script dependency filepaths.
     */
    getScripts() {
        return ['moment.js'];
    },

    /**
     * @function getTranslations
     * @description Translations for this module.
     * @override
     *
     * @returns {Object.<string, string>} Available translations for this module (key: language code, value: filepath).
     */
    getTranslations() {
        return {
            en: 'translations/en.json',
            de: 'translations/de.json',
            fr: 'translations/fr.json',
            id: 'translations/id.json',
            ko: 'translations/ko.json'
        };
    },

    /**
     * @function getTemplate
     * @description Nunjuck template.
     * @override
     *
     * @returns {string} Path to nunjuck template.
     */
    getTemplate() {
        return 'templates/MMM-AlarmClock.njk';
    },

    /**
     * @function getTemplateData
     * @description Data that gets rendered in the nunjuck template.
     * @override
     *
     * @returns {string} Data for the nunjuck template.
     */
    getTemplateData() {
        let src = this.config.sound;

        if (this.alarmFired && this.next.sound) {
            src = this.next.sound;
        }

        if (!src.match(/^https?:\/\//)) {
            src = this.file(`sounds/${src}`);
        }

        return {
            config: this.config,
            next: this.next,
            alarmFired: this.alarmFired,
            src
        };
    },

    /**
     * @function start
     * @description Sets first alarm and creates interval to check the alarm event.
     * @override
     *
     * @returns {void}
     */
    start() {
        Log.info(`Starting module: ${this.name}`);
        this.setNextAlarm();
        setInterval(() => {
            this.checkAlarm();
        }, 1000);
        moment.locale(config.language);
    },

    /**
     * @function notificationReceived
     * @description Handles incoming broadcasts from other modules or the MagicMirror core.
     * @override
     *
     * @param {string} notification - Notification name
     *
     * @returns {void}
     */
    notificationReceived(notification) {
        if (notification === 'STOP_ALARM') {
            this.resetAlarmClock();
        }
    },

    /**
     * @function checkAlarm
     * @description Checks the alarm event and triggers the alert.
     *
     * @returns {void}
     */
    checkAlarm() {
        if (!this.alarmFired && this.next && moment().diff(this.next.moment) >= 0) {
            const alert = {
                imageFA: 'bell-o',
                title: this.next.sender || this.next.title,
                message: this.next.message
            };
            let timer = this.config.timer;
            // If the alarm has specific timer and if MM is not touch, we use the alarm timer.
            if (typeof this.next.timer !== 'undefined' && !this.config.touch) {
                timer = this.next.timer;
            }
            if (!this.config.touch) {
                alert.timer = timer;
            }
            if (this.config.popup) {
                this.sendNotification('SHOW_ALERT', alert);
            }
            this.alarmFired = true;
            this.updateDom();
            this.timer = setTimeout(() => {
                this.resetAlarmClock();
            }, timer);
            setTimeout(() => {
                if (this.config.touch && this.config.popup) {
                    MM.getModules().enumerate(module => {
                        if (module.name === 'alert') {
                            module.alerts['MMM-AlarmClock'].ntf.addEventListener('click', () => {
                                this.resetAlarmClock();
                            });
                        }
                    });
                }

                const player = document.getElementById('MMM-AlarmClock-Player');
                player.volume = this.config.fade ? 0 : this.config.volume;
                this.fadeAlarm();
            }, 100);
        }
    },

    /**
     * @function fadeAlarm
     * @description Fades in the alarm sound step by step.
     *
     * @returns {void}
     */
    fadeAlarm() {
        let volume = this.config.fade ? 0 : this.config.volume;
        let counter = 0;
        const player = document.getElementById('MMM-AlarmClock-Player');
        this.fadeInterval = setInterval(() => {
            player.volume = volume;
            volume += this.config.fadeStep;
            counter += 1000;
            if (volume >= this.config.volume || counter >= this.config.fadeTimer) {
                player.volume = this.config.volume;
                clearInterval(this.fadeInterval);
            }
        }, 1000);
    },

    /**
     * @function setNextAlarm
     * @description Sets the next occurring alarm event.
     *
     * @returns {void}
     */
    setNextAlarm() {
        this.next = null;
        for (let i = 0; i < this.config.alarms.length; i += 1) {
            const temp = this.getMoment(this.config.alarms[i]);
            if (!this.next || temp.diff(this.next.moment) < 0) {
                this.next = this.config.alarms[i];
                this.next.moment = temp;
            }
        }
    },

    /**
     * @function resetAlarmClock
     * @description Resets the alarm clock.
     *
     * @returns {void}
     */
    resetAlarmClock() {
        clearTimeout(this.timer);
        clearTimeout(this.fadeInterval);
        this.alarmFired = false;
        if (this.config.touch && this.config.popup) {
            this.sendNotification('HIDE_ALERT');
        }
        this.setNextAlarm();
        this.updateDom(300);
    },

    /**
     * @function getMoment
     * @description Creates a moment object based on alarm event.
     *
     * @param {Object} alarm - Alarm event.
     * @param {string} alarm.time - Alarm event time.
     * @param {int[]} alarm.days - Weekdays the alarm event occurrs.
     *
     * @returns {Object} Moment object for the next occurrence of this alarm event.
     *
     * @example <caption>alarm object</caption>
     * {
     *      time: "18:30",
     *      days: [2,4],
     *      title: "Soccer",
     *      message: "Get ready for soccer training!"
     * }
     */
    getMoment(alarm) {
        const now = moment();
        let difference = Math.min();
        const hour = parseInt(alarm.time.split(':')[0]);
        const minute = parseInt(alarm.time.split(':')[1]);

        for (let i = 0; i < alarm.days.length; i += 1) {
            if (now.day() < alarm.days[i]) {
                difference = Math.min(alarm.days[i] - now.day(), difference);
            } else if (now.day() === alarm.days[i] && (parseInt(now.hour()) < hour
                || parseInt(now.hour()) === hour && parseInt(now.minute()) < minute)) {
                difference = Math.min(0, difference);
            } else if (now.day() === alarm.days[i]) {
                difference = Math.min(7, difference);
            } else {
                difference = Math.min(7 - now.day() + alarm.days[i], difference);
            }
        }

        return moment().add(difference, 'days')
            .set({
                hour,
                minute,
                second: 0,
                millisecond: 0
            });
    }
});
